Firestore/Source/Core/FSTView.mm

```diff
@@ -201,19 +202,40 @@ - (instancetype)initWithQuery:(FSTQuery *)query remoteDocuments:(DocumentKeySet)
   return _syncedDocuments;
 }
 
-- (FSTViewDocumentChanges *)computeChangesWithDocuments:(const MaybeDocumentMap &)docChanges {
-  return [self computeChangesWithDocuments:docChanges previousChanges:nil];
+// These two overloads allow skipping a cast when dealing with a map of `FSTDocument`s.
+FSTDocument *GetFSTDocumentOrNil(FSTMaybeDocument *maybeDoc) {
+  if ([maybeDoc isKindOfClass:[FSTDocument class]]) {
+    return static_cast<FSTDocument *>(maybeDoc);
+  }
+  return nil;
+}
+FSTDocument *GetFSTDocumentOrNil(FSTDocument *doc) {
+  return doc;
 }
 
-- (FSTViewDocumentChanges *)computeChangesWithDocuments:(const MaybeDocumentMap &)docChanges
-                                        previousChanges:
-                                            (nullable FSTViewDocumentChanges *)previousChanges {
+bool ShouldWaitForSyncedDocument(FSTDocument *newDoc, FSTDocument *oldDoc) {
+  // We suppress the initial change event for documents that were modified as part of a write
+  // acknowledgment (e.g. when the value of a server transform is applied) as Watch will send us
+  // the same document again. By suppressing the event, we only raise two user visible events (one
+  // with `hasPendingWrites` and the final state of the document) instead of three (one with
+  // `hasPendingWrites`, the modified document with `hasPendingWrites` and the final state of the
+  // document).
+  return (oldDoc.hasLocalMutations && newDoc.hasCommittedMutations && !newDoc.hasLocalMutations);
+}
+
+// Shared implementation of the `computeChanges` methods; the customization point is in
+// `GetFSTDocumentOrNil` function.
+template <typename MapT>
+FSTViewDocumentChanges *ComputeChanges(FSTView *view,
+                                       const DocumentKeySet &mutatedKeys,
+                                       const MapT &docChanges,
+                                       FSTViewDocumentChanges *previousChanges) {
   FSTDocumentViewChangeSet *changeSet =
       previousChanges ? previousChanges.changeSet : [FSTDocumentViewChangeSet changeSet];
-  FSTDocumentSet *oldDocumentSet = previousChanges ? previousChanges.documentSet : self.documentSet;
+  FSTDocumentSet *oldDocumentSet = previousChanges ? previousChanges.documentSet : view.documentSet;
 
-  DocumentKeySet newMutatedKeys = previousChanges ? previousChanges.mutatedKeys : _mutatedKeys;
-  DocumentKeySet oldMutatedKeys = _mutatedKeys;
+  DocumentKeySet newMutatedKeys = previousChanges ? previousChanges.mutatedKeys : mutatedKeys;
```

Comment:
Is mutatedKeys just self->_mutatedKeys? If so, you might be able to eliminate a parameter to this function.